app/controllers/users_controller.rb

```diff
@@ -0,0 +1,34 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.create(user_params)
+    if @user.save
+      redirect_to new_user_path
+    else
+      render :new, status: :unprocessable_entity # important to return the error message
+
+    end
+  end
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
+    if @user.update(user_params)
+      redirect_to new_user_path
+    else
+      render :edit, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit('username', 'email', 'password')
+  end
```

Comment:
- [ ] To reference the variables correctly, it will be better to write them in this format
```ruby
params.require(:user).permit(:username, :email, :password)
```